File: rake-ios-tutorial-ObjC/rake-ios-tutorial-ObjC/sdk/shuttle/DiRakeClientTestIOsSentinelShuttle.m
/*
 * Sentinel-Shuttle for Objective-C
 *
 * by Data Infrastructure
 * Template version
 * - 0.1.1 : first release
 * - 0.1.2 : log_id methods
 * - 0.1.3 : remove _$ssToken
 * - 0.1.4 : toNSDictionary(NSDictionary*)
 * - 0.1.5 : extra partition exception
 *
 * Author
 *  - Sentinel Shuttle Generator v3.0
 *  - hyeyoung.song@sk.com (Data Infrastructure Team)
 */

#import <objc/message.h>
#import <foundation/NSJSONSerialization.h>
#import "DiRakeClientTestIOsSentinelShuttle.h"



@interface DiRakeClientTestIOsSentinelShuttle()

// private members (meta data)

// private fields
@property(nonatomic) NSString * base_time;
@property(nonatomic) NSString * local_time;
@property(nonatomic) NSString * recv_time;
@property(nonatomic) NSString * os_name;
@property(nonatomic) NSString * os_version;
@property(nonatomic) NSString * resolution;
@property(nonatomic) NSNumber * screen_width;
@property(nonatomic) NSNumber * screen_height;
@property(nonatomic) NSString * language_code;
@property(nonatomic) NSString * rake_lib;
@property(nonatomic) NSString * rake_lib_version;
@property(nonatomic) NSString * ip;
@property(nonatomic) NSString * recv_host;
@property(nonatomic) NSString * token;
@property(nonatomic) NSString * log_version;
@property(nonatomic) NSString * device_id;
@property(nonatomic) NSString * device_model;
@property(nonatomic) NSString * manufacturer;
@property(nonatomic) NSString * carrier_name;
@property(nonatomic) NSString * network_type;
@property(nonatomic) NSString * app_release;
@property(nonatomic) NSString * app_build_number;
@property(nonatomic) NSString * browser_name;
@property(nonatomic) NSString * browser_version;
@property(nonatomic) NSString * referrer;
@property(nonatomic) NSString * url;
@property(nonatomic) NSString * document_title;
@property(nonatomic) NSString * page_id;
@property(nonatomic) NSString * action_id;
@property(nonatomic) NSString * dmpc;
@property(nonatomic) NSString * reserved02;
@property(nonatomic) NSString * context_id;
@property(nonatomic) NSString * context_desc;
@property(nonatomic) NSString * user_id;
@property(nonatomic) NSString * ab_test_group;


@end

@implementation DiRakeClientTestIOsSentinelShuttle



static NSString *_$ssTemplateVersion = @"0.1.5";
static NSString *_$ssVersion = @"19.07.24:3.1.0:7";
static NSString *_$ssSchemaId = @"DiRakeClientTestIOs";
static NSString *_$ssDelim = @"\t";
static NSString *_$logVersionKey = @"log_version";
static NSString *_$extraPartitionFieldKey = @"";

static NSArray *headerFieldNameList;
static NSArray *bodyFieldNameList;
static NSArray *actionKeyNameList;
static NSDictionary *bodyRule;
static NSArray *encryptedFieldsList;

+ (void)initialize {
    headerFieldNameList = @[@"base_time", @"local_time", @"recv_time", @"os_name", @"os_version", @"resolution", @"screen_width", @"screen_height", @"language_code", @"rake_lib", @"rake_lib_version", @"ip", @"recv_host", @"token", @"log_version", @"device_id", @"device_model", @"manufacturer", @"carrier_name", @"network_type", @"app_release", @"app_build_number", @"browser_name", @"browser_version", @"referrer", @"url", @"document_title", @"page_id", @"action_id", @"dmpc", @"reserved02"];
    bodyFieldNameList = @[@"context_id", @"context_desc", @"user_id", @"ab_test_group"];
    actionKeyNameList = @[@"page_id", @"action_id"];
    encryptedFieldsList = @[@"device_id"];

    bodyRule = [[NSMutableDictionary alloc]init];



}

- (id)init {
    self = [super init];
    if (self) {
        _log_version = _$ssVersion;
    }

    return self;
}


/**
 * Methods
 */

- (DiRakeClientTestIOsSentinelShuttle *) base_time:(NSString *) base_time {
    _base_time = base_time;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) local_time:(NSString *) local_time {
    _local_time = local_time;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) recv_time:(NSString *) recv_time {
    _recv_time = recv_time;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) os_name:(NSString *) os_name {
    _os_name = os_name;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) os_version:(NSString *) os_version {
    _os_version = os_version;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) resolution:(NSString *) resolution {
    _resolution = resolution;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) screen_width:(NSNumber *) screen_width {
    _screen_width = screen_width;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) screen_height:(NSNumber *) screen_height {
    _screen_height = screen_height;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) language_code:(NSString *) language_code {
    _language_code = language_code;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) rake_lib:(NSString *) rake_lib {
    _rake_lib = rake_lib;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) rake_lib_version:(NSString *) rake_lib_version {
    _rake_lib_version = rake_lib_version;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) ip:(NSString *) ip {
    _ip = ip;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) recv_host:(NSString *) recv_host {
    _recv_host = recv_host;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) token:(NSString *) token {
    _token = token;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) device_id:(NSString *) device_id {
    _device_id = device_id;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) device_model:(NSString *) device_model {
    _device_model = device_model;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) manufacturer:(NSString *) manufacturer {
    _manufacturer = manufacturer;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) carrier_name:(NSString *) carrier_name {
    _carrier_name = carrier_name;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) network_type:(NSString *) network_type {
    _network_type = network_type;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) app_release:(NSString *) app_release {
    _app_release = app_release;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) app_build_number:(NSString *) app_build_number {
    _app_build_number = app_build_number;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) browser_name:(NSString *) browser_name {
    _browser_name = browser_name;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) browser_version:(NSString *) browser_version {
    _browser_version = browser_version;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) referrer:(NSString *) referrer {
    _referrer = referrer;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) url:(NSString *) url {
    _url = url;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) document_title:(NSString *) document_title {
    _document_title = document_title;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) page_id:(NSString *) page_id {
    _page_id = page_id;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) action_id:(NSString *) action_id {
    _action_id = action_id;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) dmpc:(NSString *) dmpc {
    _dmpc = dmpc;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) reserved02:(NSString *) reserved02 {
    _reserved02 = reserved02;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) context_id:(NSString *) context_id {
    _context_id = context_id;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) context_desc:(NSString *) context_desc {
    _context_desc = context_desc;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) user_id:(NSString *) user_id {
    _user_id = user_id;
    return self;
}

- (DiRakeClientTestIOsSentinelShuttle *) ab_test_group:(NSString *) ab_test_group {
    _ab_test_group = ab_test_group;
    return self;
}
- (DiRakeClientTestIOsSentinelShuttle *) setBodyOfpage1__action1_with_context_id:(NSString *) context_id
context_desc:(NSString *) context_desc
 {
    [self clearBody];
    _page_id = @"page1";
    _action_id = @"action1";
    _context_id = context_id;
    _context_desc = context_desc;

    return self;
}




// 12 public util functions
- (NSString *) toString {
    return [self toHBString];
}

- (NSString *) toHBString {
    return [self toHBString: _$ssDelim];
}

- (NSString *) toHBString:(NSString *)delim {
    return [NSString stringWithFormat:@"%@%@",[self headerToString:delim],[self bodyToString]];
}

- (NSString *) headerToString:(NSString *)delim {
    NSString *headerString = @"";
    for (NSString *fieldName in headerFieldNameList) {

        NSString *valueString = @"";
        NSObject *value = [self valueForKey:fieldName];

        if (value != nil) {
            if ([value isKindOfClass:[NSNumber class]]) {
                valueString = [NSString stringWithFormat:@"%@",value];
            } else if ([value isKindOfClass:[NSString class]]) {
                valueString = [self getEscapedValue:(NSString*)value];
            } else if ([value isKindOfClass:[NSMutableDictionary class]]) {
                valueString = [self convertFromNSMutableDictionaryToString:(NSMutableDictionary *)value];
            }
        } else {
            if ([_$extraPartitionFieldKey isEqualToString:fieldName]) {
                @throw [NSException exceptionWithName:@"IllegalArgumentException"
                                               reason:[NSString stringWithFormat:@"%@ should not be empty.", fieldName]
                                             userInfo: nil];
        }
     }

        headerString = [headerString stringByAppendingString:[valueString stringByAppendingString:delim]];
    }

    return headerString;
}

- (void) clearBody {
    for (NSString *bodyFieldName in bodyFieldNameList) {
        [self setValue:nil forKey:bodyFieldName];
    }
    
}


- (NSDictionary *) getBody {
    
    NSMutableDictionary *body = [[NSMutableDictionary alloc] init];

    for (NSString *bodyFieldName in bodyFieldNameList) {
        if ([self valueForKey:bodyFieldName]) {
            [body setValue:[self valueForKey:bodyFieldName] forKey:bodyFieldName];
        }
    }

    return body;
}

- (NSString *) bodyToString {
    NSError *err;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:[self getBody] options:0 error:&err];
    NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

    return [self getEscapedValue:jsonString];
}

+ (NSDictionary *) getSentinelMeta {

    NSMutableDictionary *fieldOrder = [[NSMutableDictionary alloc] init];

    int i = 0;
    for (i = 0; i < headerFieldNameList.count ; i++) {
        NSString *headerFieldName = [headerFieldNameList objectAtIndex:i];
        [fieldOrder setValue:[NSNumber numberWithInt:i] forKey:headerFieldName];
    }

    [fieldOrder setValue:[NSNumber numberWithInt:i] forKey:@"_$body"];

    NSDictionary *sentinel_meta = @{@"sentinel_meta":@{
                                            @"_$schemaId": _$ssSchemaId,
                                            @"_$fieldOrder":fieldOrder,
                                            @"_$encryptionFields":encryptedFieldsList,
                                            @"_$projectId":@""
                                            }};

    return sentinel_meta;
}

+ (NSDictionary *) toNSDictionary:(NSDictionary *)dict {
    NSMutableDictionary *sentinelDictionary = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *bodyDictionary = [[NSMutableDictionary alloc] init];

    if (![_$extraPartitionFieldKey isEqualToString:@""]) {
        if (![[dict allKeys] containsObject:_$extraPartitionFieldKey]) {
            @throw [NSException exceptionWithName:@"IllegalArgumentException"
                                           reason:[NSString stringWithFormat:@"%@ should be exist.", _$extraPartitionFieldKey]
                                         userInfo: nil];
        }
    }

    for (NSString *fieldName in headerFieldNameList) {
        if ([dict valueForKey:fieldName] != nil) {
            sentinelDictionary[fieldName] = dict[fieldName];
        } else {
            sentinelDictionary[fieldName] = @"";
        }
    }

    sentinelDictionary[_$logVersionKey] = _$ssVersion;

    for (NSString* fieldName in bodyFieldNameList) {
        if ([dict valueForKey:fieldName] != nil) {
            bodyDictionary[fieldName] = dict[fieldName];
        }
    }

    [sentinelDictionary setValue:bodyDictionary forKey:@"_$body"];
    [sentinelDictionary setValue:[[DiRakeClientTestIOsSentinelShuttle getSentinelMeta] valueForKey:@"sentinel_meta"] forKey:@"sentinel_meta"];

    return sentinelDictionary;
}

- (NSDictionary *) toNSDictionary {
    NSMutableDictionary* properties = [[NSMutableDictionary alloc] init];

    // header
    for (NSString *headerFieldName in headerFieldNameList) {
        NSString *valueString = @"";
        if ([self valueForKey:headerFieldName] != nil) {
            valueString = [self valueForKey:headerFieldName];
            if ([valueString isKindOfClass:[NSString class]]) {
                valueString = [self getEscapedValue:valueString];
            }
        } else {
            if ([_$extraPartitionFieldKey isEqualToString:headerFieldName]) {
                @throw [NSException exceptionWithName:@"IllegalArgumentException"
                                               reason:[NSString stringWithFormat:@"%@ should not be empty.", headerFieldName]
                                             userInfo: nil];
            }
        }
        [properties setValue:valueString forKey:headerFieldName];
    }

    // body
    [properties setValue:[self getBody] forKey:@"_$body"];

    // sentinel_meta
    [properties setValue:[[DiRakeClientTestIOsSentinelShuttle getSentinelMeta] valueForKey:@"sentinel_meta"] forKey:@"sentinel_meta"];

    return properties;
}


- (NSString *) toJSONString {
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:[self toNSDictionary]
                                                       options:(NSJSONWritingOptions)NSJSONWritingPrettyPrinted
                                                       error:&error];
    if (! jsonData) {
        NSLog(@"bv_jsonStringWithPrettyPrint: error: %@", error.localizedDescription);
        return @"{}";
    } else {
        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
}

- (NSString *) getEscapedValue:(NSString *)value {
    return [[[value stringByReplacingOccurrencesOfString:@"\n" withString:@"\\n"] stringByReplacingOccurrencesOfString:@"\t" withString:@"\\t"] stringByReplacingOccurrencesOfString:@"\r" withString:@"\\r"];
}

- (NSString *) convertFromNSMutableDictionaryToString: (NSMutableDictionary *)dict {
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dict options:0 error:nil];
    NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    return jsonString;
}

@end

File: rake-ios-tutorial-Swift-Storyboard/rake-ios-tutorial-Swift-Storyboard/sdk/shuttle/DiRakeClientTestIOsSentinelShuttle.swift
/*
 * Sentinel-Shuttle for Swift
 *
 * by Data Infrastructure
 * Template version
 * - 0.1.1 : first release
 * - 0.1.2 : log_id methods
 * - 0.1.3 : remove _$ssToken
 * - 0.1.4 : toDictionary() -> [String:Any]
 * - 0.1.5 : extra partition exception
 *
 * Author
 *  - Sentinel Shuttle Generator v3.0
 *  - hyeyoung.song@sk.com (Data Infrastructure Team)
 */

import Foundation

public class DiRakeClientTestIOsSentinelShuttle {
    public enum DiRakeClientTestIOsSentinelShuttleError: Error {
        case EmptyExtraPartitionFieldError(reason:String)
    }

    private static let _$ssTemplateVersion = "0.1.5"
    private static let _$ssVersion = "22.08.22:4.0.5:11"
    private static let _$ssSchemaId = "DiRakeClientTestIOs"
    private static let _$ssDelim = "\t"
    private static let _$logVersionKey = "log_version"
    private static let _$extraPartitionFieldKey = ""

    private static let headerFieldNameList:[String] = ["base_time", "local_time", "recv_time", "os_name", "os_version", "resolution", "screen_width", "screen_height", "language_code", "rake_lib", "rake_lib_version", "ip", "recv_host", "token", "log_version", "device_id", "device_model", "manufacturer", "carrier_name", "network_type", "app_release", "app_build_number", "browser_name", "browser_version", "referrer", "url", "document_title", "page_id", "action_id", "dmpc", "reserved02"]
    private static let bodyFieldNameList:[String] = ["context_id", "context_desc", "user_id", "ab_test_group"]
    private static let actionKeyNameList:[String] = ["page_id", "action_id"]
    //    private static var bodyRule = [:]   // String,Any?
    private static let encryptedFieldsList:[String] = ["device_id"]

    private var fieldsDictionary:[String:Any?] = [:]
    

    public init() {

        for headerFieldName in DiRakeClientTestIOsSentinelShuttle.headerFieldNameList {
            if headerFieldName == DiRakeClientTestIOsSentinelShuttle._$logVersionKey {
                self.fieldsDictionary.updateValue(DiRakeClientTestIOsSentinelShuttle._$ssVersion, forKey: DiRakeClientTestIOsSentinelShuttle._$logVersionKey)
            } else {
                self.fieldsDictionary.updateValue(nil, forKey: headerFieldName)
            }
        }

        for bodyFieldName in DiRakeClientTestIOsSentinelShuttle.bodyFieldNameList {
            self.fieldsDictionary.updateValue(nil, forKey: bodyFieldName)
        }
    }

    /*
     Methods
     */

    public func base_time(_ base_time:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(base_time, forKey: "base_time")
        return self
    }

    public func local_time(_ local_time:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(local_time, forKey: "local_time")
        return self
    }

    public func recv_time(_ recv_time:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(recv_time, forKey: "recv_time")
        return self
    }

    public func os_name(_ os_name:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(os_name, forKey: "os_name")
        return self
    }

    public func os_version(_ os_version:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(os_version, forKey: "os_version")
        return self
    }

    public func resolution(_ resolution:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(resolution, forKey: "resolution")
        return self
    }

    public func screen_width(_ screen_width:Int?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(screen_width, forKey: "screen_width")
        return self
    }

    public func screen_height(_ screen_height:Int?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(screen_height, forKey: "screen_height")
        return self
    }

    public func language_code(_ language_code:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(language_code, forKey: "language_code")
        return self
    }

    public func rake_lib(_ rake_lib:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(rake_lib, forKey: "rake_lib")
        return self
    }

    public func rake_lib_version(_ rake_lib_version:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(rake_lib_version, forKey: "rake_lib_version")
        return self
    }

    public func ip(_ ip:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(ip, forKey: "ip")
        return self
    }

    public func recv_host(_ recv_host:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(recv_host, forKey: "recv_host")
        return self
    }

    public func token(_ token:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(token, forKey: "token")
        return self
    }

    public func device_id(_ device_id:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(device_id, forKey: "device_id")
        return self
    }

    public func device_model(_ device_model:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(device_model, forKey: "device_model")
        return self
    }

    public func manufacturer(_ manufacturer:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(manufacturer, forKey: "manufacturer")
        return self
    }

    public func carrier_name(_ carrier_name:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(carrier_name, forKey: "carrier_name")
        return self
    }

    public func network_type(_ network_type:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(network_type, forKey: "network_type")
        return self
    }

    public func app_release(_ app_release:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(app_release, forKey: "app_release")
        return self
    }

    public func app_build_number(_ app_build_number:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(app_build_number, forKey: "app_build_number")
        return self
    }

    public func browser_name(_ browser_name:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(browser_name, forKey: "browser_name")
        return self
    }

    public func browser_version(_ browser_version:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(browser_version, forKey: "browser_version")
        return self
    }

    public func referrer(_ referrer:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(referrer, forKey: "referrer")
        return self
    }

    public func url(_ url:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(url, forKey: "url")
        return self
    }

    public func document_title(_ document_title:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(document_title, forKey: "document_title")
        return self
    }

    public func page_id(_ page_id:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(page_id, forKey: "page_id")
        return self
    }

    public func action_id(_ action_id:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(action_id, forKey: "action_id")
        return self
    }

    public func dmpc(_ dmpc:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(dmpc, forKey: "dmpc")
        return self
    }

    public func reserved02(_ reserved02:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(reserved02, forKey: "reserved02")
        return self
    }

    public func context_id(_ context_id:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(context_id, forKey: "context_id")
        return self
    }

    public func context_desc(_ context_desc:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(context_desc, forKey: "context_desc")
        return self
    }

    public func user_id(_ user_id:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(user_id, forKey: "user_id")
        return self
    }

    public func ab_test_group(_ ab_test_group:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.fieldsDictionary.updateValue(ab_test_group, forKey: "ab_test_group")
        return self
    }

    public func setBodyOfpage1__action1(context_id:String?, context_desc:String?) -> DiRakeClientTestIOsSentinelShuttle {
        self.clearBody()
        self.fieldsDictionary.updateValue("page1", forKey: "page_id")
        self.fieldsDictionary.updateValue("action1", forKey: "action_id")
        self.fieldsDictionary.updateValue(context_id, forKey: "context_id")
        self.fieldsDictionary.updateValue(context_desc, forKey: "context_desc")

        return self
    }

    public func setBodyOftest__test() -> DiRakeClientTestIOsSentinelShuttle {
        self.clearBody()
        self.fieldsDictionary.updateValue("test", forKey: "page_id")
        self.fieldsDictionary.updateValue("test", forKey: "action_id")

        return self
    }




    /*
     12 public util functions
     */
    public func toString() -> String {
        return self.toHBString()
    }

    public func toHBString() -> String {
        return self.toHBString(DiRakeClientTestIOsSentinelShuttle._$ssDelim)
    }

    public func toHBString(_ delim:String) -> String {
        do {
            return try "\(self.headerToString(delim))\(self.bodyToString())"
        } catch let error {
            print("\(self) \(error)");
            return ""
        }
    }

    public func headerToString(_ delim:String) throws -> String {
        var headerString = ""

        for fieldName in DiRakeClientTestIOsSentinelShuttle.headerFieldNameList {
            var valueString = ""

            if let value = self.fieldsDictionary[fieldName] {
                if value is String {
                    valueString = self.getEscapedValue(value as! String)
                } else if value is [AnyHashable:Any] {
                    valueString = self.convertFromDictionaryToString(value as! [AnyHashable : Any])
                } else {
                    valueString = String(describing: value) //FIXME is it right?
                }
            } else {
                // extra partition field는 빈값이어서는 안된다.
                if DiRakeClientTestIOsSentinelShuttle._$extraPartitionFieldKey == fieldName {
                    throw DiRakeClientTestIOsSentinelShuttleError.EmptyExtraPartitionFieldError(reason: "\(self) \(fieldName) should not be empty.")
                }
            }

            headerString = headerString.appending("\(valueString)\(delim)")
        }

        return headerString
    }

    public func clearBody() {
        for bodyFieldName in DiRakeClientTestIOsSentinelShuttle.bodyFieldNameList {
            self.fieldsDictionary.updateValue(nil, forKey: bodyFieldName)
        }
        
    }

    public func getBody() -> [String:Any] {

        var body:[String:Any] = [:]

        for bodyFieldName in DiRakeClientTestIOsSentinelShuttle.bodyFieldNameList {
            let bodyFieldValue = self.fieldsDictionary[bodyFieldName]!
            if bodyFieldValue != nil {
                body.updateValue(bodyFieldValue!, forKey: bodyFieldName)
            }
        }

        return body
    }

    public func bodyToString() -> String {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: self.getBody(), options: JSONSerialization.WritingOptions(rawValue: 0))
            let jsonString =  String.init(data: jsonData, encoding: String.Encoding.utf8)!
            return self.getEscapedValue(jsonString)
        } catch let error {
            print("\(self) bodyToString() failed: \(error)")
            return "{}"
        }
    }

    public static func getSentinelMeta() -> [String:[String:Any]] {
        var fieldOrder:[String:Int] = [:]
        for (index, headerFieldName) in DiRakeClientTestIOsSentinelShuttle.headerFieldNameList.enumerated() {
            fieldOrder.updateValue(index, forKey: headerFieldName)
        }

        fieldOrder.updateValue(DiRakeClientTestIOsSentinelShuttle.headerFieldNameList.count, forKey: "_$body")

        return [
            "sentinel_meta": [
                "_$schemaId": DiRakeClientTestIOsSentinelShuttle._$ssSchemaId,
                "_$fieldOrder": fieldOrder,
                "_$encryptionFields": DiRakeClientTestIOsSentinelShuttle.encryptedFieldsList,
                "_$projectId": ""
            ]
        ]
    }

    public static func toDictionary(_ dict:[String:Any?]) throws -> [String:Any] {
        var sentinelDictionary:[String:Any] = [:]
        var bodyDictionary:[String:Any] = [:]

        if DiRakeClientTestIOsSentinelShuttle._$extraPartitionFieldKey != "" {
            if !dict.keys.contains(DiRakeClientTestIOsSentinelShuttle._$extraPartitionFieldKey) {
                throw DiRakeClientTestIOsSentinelShuttleError.EmptyExtraPartitionFieldError(reason: "\(self) \(DiRakeClientTestIOsSentinelShuttle._$extraPartitionFieldKey) should be exist.")
            }
        }

        for fieldName in DiRakeClientTestIOsSentinelShuttle.headerFieldNameList {
            if let value = dict[fieldName] {
                sentinelDictionary.updateValue(value!, forKey: fieldName)
            } else {
                sentinelDictionary.updateValue("", forKey: fieldName)
            }
        }

        sentinelDictionary.updateValue(DiRakeClientTestIOsSentinelShuttle._$ssVersion, forKey: DiRakeClientTestIOsSentinelShuttle._$logVersionKey)

        for fieldName in DiRakeClientTestIOsSentinelShuttle.bodyFieldNameList {
            if let value = dict[fieldName] {
                bodyDictionary.updateValue(value!, forKey: fieldName)
            }
        }

        sentinelDictionary.updateValue(bodyDictionary, forKey: "_$body")
        sentinelDictionary.updateValue(self.getSentinelMeta()["sentinel_meta"]!, forKey: "sentinel_meta")

        return sentinelDictionary
    }

    public func toDictionary() throws -> [String:Any] {
        var properties:[String:Any] = [:]

        // header
        for headerFieldName  in DiRakeClientTestIOsSentinelShuttle.headerFieldNameList {
            var valueString = ""

            if let value = self.fieldsDictionary[headerFieldName]! {
                if value is String {
                    valueString = self.getEscapedValue(value as! String)
                } else {
                    valueString = String(describing: value)
                }
            } else {
                if DiRakeClientTestIOsSentinelShuttle._$extraPartitionFieldKey == headerFieldName {
                    throw DiRakeClientTestIOsSentinelShuttleError.EmptyExtraPartitionFieldError(reason: "\(self) \(headerFieldName) should not be empty.")
                }
            }

            properties.updateValue(valueString, forKey: headerFieldName)
        }

        // body
        properties.updateValue(self.getBody(), forKey: "_$body")

        // sentinel_meta
        properties.updateValue(DiRakeClientTestIOsSentinelShuttle.getSentinelMeta()["sentinel_meta"]!, forKey: "sentinel_meta")

        return properties
    }

    public func toJSONString() -> String {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: self.toDictionary(), options: JSONSerialization.WritingOptions.prettyPrinted)
            return String.init(data: jsonData, encoding: String.Encoding.utf8)!
        } catch let error {
            print("\(self) toJSONString() failed: \(error)")
            return "{}"
        }
    }

    public func getEscapedValue(_ value:String) -> String {
        return value.replacingOccurrences(of: "\n", with: "\\n").replacingOccurrences(of: "\t", with: "\\t").replacingOccurrences(of: "\r", with: "\\r")
    }

    func convertFromDictionaryToString(_ dict:[AnyHashable:Any]) -> String {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dict, options: JSONSerialization.WritingOptions(rawValue: 0))
            return String.init(data: jsonData, encoding: String.Encoding.utf8)!
        } catch let error {
            print("\(self) convertFromDictionaryToString() failed: \(error)")
            return "{}"
        }
    }
}
